babel.config.js

```diff
@@ -25,99 +25,394 @@
+  const postClosurePlugins = argv.esm
+    ? [
+        localPlugin('transform-minified-comments'),
+        localPlugin('transform-function-declarations'),
+        localPlugin('transform-stringish-literals'),
+      ]
+    : [];
+  return {
+    inputSourceMap: false,
+    plugins: postClosurePlugins,
+    retainLines: false,
+    sourceMaps: true,
+    sourceType: 'module',
+  };
+}
 
-const presets = (esm) => {
-  if (esm) {
-    return [
-      [
-        '@babel/preset-env',
-        {
-          'modules': false,
-          'targets': targets(esm),
-          'bugfixes': true,
-        },
-      ],
-    ];
-  }
-
-  return [
-    [
-      '@babel/preset-env',
-      {
-        'modules': isClosureCompiler ? false : 'commonjs',
-        'loose': true,
-        'targets': targets(esm),
-      },
-    ],
+/**
+ * Gets the config for babel transforms run during `gulp dist --single_pass`.
+ *
+ * @return {!Object}
+ */
+function getSinglePassConfig() {
+  const singlePassPlugins = [localPlugin('transform-prune-namespace')];
+  return {
+    compact: false,
+    inputSourceMap: false,
+    plugins: singlePassPlugins,
+    sourceMaps: true,
+  };
+}
+
+/**
+ * Gets the config for babel transforms run during `gulp [unit|integration]`.
+ *
+ * @return {!Object}
+ */
+function getTestConfig() {
+  const testPresets = [defaultPresetEnvPlugin];
+  const testPlugins = [
+    replacePlugin,
+    localPlugin('transform-json-configuration'),
+    localPlugin('transform-fix-leading-comments'),
+    babelPlugin('transform-react-constant-elements'),
+    [babelPlugin('transform-classes'), {loose: false}],
+    [babelPlugin('transform-react-jsx'), defaultJsxOpts],
   ];
-};
+  if (argv.converage) {
+    testPlugins.unshift(defaultInstanbulPlugin);
+  }
+  return {
+    compact: false,
+    ignore: ignoredGlobalModules,
+    plugins: testPlugins,
+    presets: testPresets,
+    sourceType: 'module',
+  };
+}
 
-// eslint-disable-next-line local/no-module-exports
+/**
+ * Main entry point. Returns different babel configs based on the caller.
+ *
+ * @param {!Object} api
+ * @return {!Object}
+ */
 module.exports = function (api) {
-  api.cache(true);
-  // Closure Compiler builds do not use any of the default settings below until its
-  // an esm build. (Both Multipass and Singlepass)
-  if (isClosureCompiler && !esm) {
-    return {};
+  /**
+   * Identifies the caller of a babel transform.
+   *
+   * @param {string} name
+   * @return {boolean}
+   */
+  function calledBy(name) {
+    return api.caller((caller) => !!(caller && caller.name === name));
   }
 
-  return {
-    'plugins': plugins(esm),
-    'presets': presets(esm),
-    'compact': false,
-    'sourceType': 'module',
-  };
+  /**
+   * Throws an error when an unrecognized mode is encountered.
+   */
+  function throwError() {
+    const err = new Error('Unrecognized Babel mode.');
+    err.showStack = false;
+    throw err;
+  }
+
+  if (calledBy('dep-check')) {
```

Comment:
This is a bit strange to me. This looks like a `Map` hiding inside.

```javascript
if (configs.has(api.caller(....)) {
  return configs.get(api.caller(...));
}
```

Reply:
Agree. Changed this to a map.